File: src/main/java/quebec/virtualite/backend/services/rest/impl/RestServer.java
package quebec.virtualite.backend.services.rest.impl;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;
import quebec.virtualite.backend.services.domain.DomainService;
import quebec.virtualite.backend.services.domain.entities.CityAlreadyExistsException;
import quebec.virtualite.backend.services.domain.entities.CityEntity;
import quebec.virtualite.backend.services.rest.CityDTO;
import quebec.virtualite.backend.services.rest.RestServerContract;

import java.util.List;

import static org.springframework.http.HttpStatus.CONFLICT;
import static org.springframework.http.HttpStatus.CREATED;
import static org.springframework.http.HttpStatus.NOT_FOUND;
import static quebec.virtualite.utils.CollectionUtils.transform;

@RestController
@RequiredArgsConstructor
@Slf4j
@Validated
public class RestServer implements RestServerContract
{
    private final DomainService domainService;

    @Override
    @PutMapping(URL_ADD_CITY__PUT)
    @ResponseStatus(CREATED)
    public void addCity(@RequestBody CityDTO city)
    {
        domainService.addCity(city.toEntity(0));
    }

    @Override
    @DeleteMapping(URL_DELETE_CITY)
    public void deleteCity(@PathVariable String name)
    {
        domainService.deleteCity(getCity(name));
    }

    @Override
    @GetMapping(URL_GET_CITY)
    public CityDTO getCityDetails(@PathVariable String name)
    {
        return new CityDTO(getCity(name));
    }

    @Override
    @GetMapping(URL_GET_CITIES)
    public List<CityDTO> getCitiesDetails()
    {
        return transform(domainService.getCities(), CityDTO::new);
    }

    @Override
    @PostMapping(URL_UPDATE_CITY__POST)
    public void updateCity(@PathVariable String name, @RequestBody CityDTO city)
    {
        CityEntity existingCity = getCity(name);
        CityEntity updatedCity = city.toEntity(existingCity.id());

        domainService.updateCity(updatedCity);
    }

    @ExceptionHandler(CityAlreadyExistsException.class)
    protected ResponseEntity<String> exceptionHandlerWhenNotFound()
    {
        return new ResponseEntity<>(CONFLICT);
    }

    private CityEntity getCity(String name)
    {
        return domainService.getCity(name)
            .orElseThrow(() -> new ResponseStatusException(NOT_FOUND));
    }
}

File: src/main/java/quebec/virtualite/backend/services/rest/impl/RestServer.kt
package quebec.virtualite.backend.services.rest.impl

import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus.CONFLICT
import org.springframework.http.HttpStatus.CREATED
import org.springframework.http.HttpStatus.NOT_FOUND
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException
import quebec.virtualite.backend.services.domain.CityAlreadyExistsException
import quebec.virtualite.backend.services.domain.DomainService
import quebec.virtualite.backend.services.domain.entities.CityEntity
import quebec.virtualite.backend.services.rest.CityDTO
import quebec.virtualite.backend.services.rest.RestServerContract
import quebec.virtualite.backend.services.rest.URL_ADD_CITY__PUT
import quebec.virtualite.backend.services.rest.URL_DELETE_CITY
import quebec.virtualite.backend.services.rest.URL_GET_CITIES
import quebec.virtualite.backend.services.rest.URL_GET_CITY
import quebec.virtualite.backend.services.rest.URL_UPDATE_CITY__POST
import quebec.virtualite.utils.CollectionUtils.map

@RestController
@Validated
open class RestServer(
    private val domainService: DomainService

) : RestServerContract
{
    private val log = LoggerFactory.getLogger(this.javaClass)

    @PutMapping(URL_ADD_CITY__PUT)
    @ResponseStatus(CREATED)
    override fun addCity(@RequestBody city: CityDTO)
    {
        domainService.addCity(city.toEntity(0))
    }

    @DeleteMapping(URL_DELETE_CITY)
    override fun deleteCity(@PathVariable name: String)
    {
        getCity(name)
        domainService.deleteCity(name)
    }

    @GetMapping(URL_GET_CITY)
    override fun getCityDetails(@PathVariable name: String): CityDTO
    {
        return CityDTO(getCity(name))
    }

    @GetMapping(URL_GET_CITIES)
    override fun getCitiesDetails(): List<CityDTO>
    {
        return map(domainService.getCitiesDetails())
        { CityDTO(it) }
    }

    @PostMapping(URL_UPDATE_CITY__POST)
    override fun updateCity(
        @PathVariable name: String,
        @RequestBody updatedCity: CityDTO
    )
    {
        domainService.updateCity(
            updatedCity.toEntity(getCity(name).id)
        )
    }

    @ExceptionHandler(CityAlreadyExistsException::class)
    internal fun exceptionHandler(e: Exception?): ResponseEntity<String>
    {
        return ResponseEntity(CONFLICT)
    }

    private fun getCity(name: String): CityEntity
    {
        return domainService.getCityDetails(name)
            ?: throw ResponseStatusException(NOT_FOUND)
    }
}
